src/single_random_draw.rs

```diff
@@ -0,0 +1,488 @@
+        None => Err(Error::Multiplication(weight, fee_rate)),
+    }
+}
+
+/// Randomly select coins for the given target by shuffling the UTXO pool and
+/// taking UTXOs until the given target is reached.
+///
+/// The fee_rate can have an impact on the selection process since the fee
+/// must be paid for in addition to the target.  However, the total fee
+/// is dependant on the number of UTXOs consumed and the new inputs created.
+/// The selection strategy therefore calculates the fees of what is known
+/// ahead of time (the number of UTXOs create and the transaction header),
+/// and then then for each new input, the effective_value is tracked which
+/// deducts the fee for each individual input at selection time.  For more
+/// discussion see the following:
+///
+/// https://bitcoin.stackexchange.com/questions/103654/calculating-fee-based-on-fee-rate-for-bitcoin-transaction/114847#114847
+///
+/// ## Parameters
+/// ///
+/// /// * `target` - target value to send to recipient.
+/// /// * `fee_rate` - ratio of transaction amount per size.
+/// /// * `utxos` - UTXOs from which to sum the target amount.
+/// /// * `inputs` - the order corresponds to the ordering of the UTXOs where the first input consumes the first UTXO and so on.
+/// /// * `transaction` - the transaction to populate with resulting inputs and outputs.
+/// /// * `script_pubkey` - needed to compose the transaction outputs.
+/// /// * `rng` - used primarily by tests to make the selection deterministic.
+///
+pub fn select_coins_srd<R: rand::Rng + ?Sized>(
+    target: Amount,
+    fee_rate: FeeRate,
+    utxos: &mut [TxOut],
+    inputs: &mut [TxIn],
+    transaction: &mut Transaction,
+    script_pubkey: ScriptBuf,
+    rng: &mut R,
+) -> Result<Option<()>, Error> {
+    utxos.shuffle(rng);
+
+    if inputs.len() != utxos.len() {
+        return Err(Error::SizeMismatch(inputs.len(), utxos.len()));
+    }
+
+    let zipped = utxos.iter().zip(inputs.iter());
+
+    let fee = calculate_fee(&script_pubkey, fee_rate, transaction)?;
+    let threshold = target + CHANGE_LOWER + fee;
+    let mut value = Amount::ZERO;
+
+    for (utxo, input) in zipped {
+
+        // note: I'd prefer to use filter() and filter all inputs that have a negative
+        //       effective_value, however, an error message is returned if a multiplication
+        //       overflow occurs while the caclulating effective_value.
+        let effective_value: Option<Amount> = get_effective_value(utxo, input, fee_rate)?;
+
+        // skip if effective_value is negative.
+        //
+        match effective_value {
+            Some(e) => value += e,
+            None => continue,
+        }
+
+        transaction.input.push(input.clone());
+
+        if transaction.weight() >= MAX_TRANSACTION_SIZE {
+            return Err(Error::MaxTransactionWeight(transaction.weight()));
+        }
+
+        if value >= threshold {
+            let change_amount = target - fee;
+
+            let change_output: TxOut =
```

Comment:
That should be possible since:

`let threshold = target + CHANGE_LOWER + fee;`

Threshold is greater than the target by change_lower, so any selection exceeds the target by at least the dust limit.

maybe it's worth putting an assert in to make sure?

Reply:
Ah I had forgotten that you wanted this to always use a change output because it's a fallback. I still don't really see why not handle the case where you just end up lucky and don't need a change output but not a big deal.

Reply:
It's probably worth adding the case where no change output is required, that way it's more modular and doesn't depend on BnB or another algo to be run first.

Reply:
this comment is outdated since I've removed the transaction building.  Therefore change_outputs are not created here (or lack of change outputs).  marking resolved.